### ISSUE TITLE
X-Frame-Options header cannot be disabled

### ISSUE THREAD

**Issue by maderlock** [magento/magento2#8504]
### Preconditions
Magento 2.1.3

### Steps to reproduce
Edit env.php to include line:
'x-frame-options' => null

### Expected result

1. Magento would not output an X-Frame-Options header

### Actual result

1. Magento outputs X-Frame-Options: SAMEORIGIN

We need a more nuanced approach to X-Frame-Options. Luckily, we can handle this within the server settings. All we need to do is stop Magento from outputting this header itself. The example nginx configuration file shows nginx handling this itself, so that suggests that there should be a way to stop Magento outputting this. Is there a different value from null that can be used?

Setting the option in env.php to blank or false still outputs the header, but with no value!

**Comment by maderlock** [comment]
Please spare me the following: "Magento always outputs a X-Frame-Options. This is for security to avoid link-jacking. You can choose between DENY, SAMEORIGIN etc."

I understand the logic behind using the header. I do not disagree.

**Comment by avoelkl** [comment]
Hi @maderlock,
Do you think it is enough if the output of the X-Frame Options header can disabled or is there also the need to set the `ALLOW-FROM` Directive as well?

`X-Frame-Options: ALLOW-FROM https://example.com/`.

The biggest risk I see is people disabling the 'X-Frame-Options' header by accident because they copy env.php settings from anywhere on the internet without knowing what it does.

**Comment by maderlock** [comment]
I think it's enough to be able to disable it, because headers like this can and should be managed by the server.

I would have thought the chance of anyone copying env.php are pretty small, as most people will not have problems with the default SAMEORIGIN. If you are concerned then the disabling value could be VERY_DANGEROUS_DISABLE_YES_I_AM_SURE :)

**Comment by avoelkl** [comment]
Pinging @piotrekkaminski to join the discussion.

**Comment by piotrekkaminski** [comment]
@maderlock what is the use case?

**Comment by nicksaroki** [comment]
@piotrekkaminski what if you need to allow multiple origin urls? that's a potential use case, right?

**Comment by maderlock** [comment]
The use case is when you need to include some pages of the frontend in an iframe on the a different domain. We use this in a number of ways - to allow some admin functionality to be displayed over the frontend pages, and to display previews of the frontend in the admin. All pretty standard stuff for  some CMS systems. However, Magento supports having different domains for the admin and frontend, so we end up having to display an iframe from different domain.

**Comment by nei** [comment]
Is there a way I could change this for just one Controller? I want to keep SAMEORIGIN for all except a controller that I am exposing to be used inside an iframe.

**Comment by michaelstephenson** [comment]
I agree with @maderlock. We should be able to disable the **X-Frame-Options** header (and a couple more like **X-Content-Type-Options** and **X-XSS-Protection**) within this file (env.php) OR within the application itself. These headers are most-often set at the server level and, from a Magento admin perspective, we should be able to stop Magento from outputting these headers if necessary. Consider the case when Magento 2 is running on a shared server (NGINX or Apache) where other web applications need the security headers set by the server. It is much nicer to be able to set them globally in the nginx.conf rather than visit each enabled-sites .conf file to include them in each server { block } as needed.

Perhaps we can see some support from Magento in the future for something like this in the env.php:

```
'x-frame-Options' => 'SAMEORIGIN',
'x-content-type-options' => 'nosniff',
'x-xss-protection' => '"1; mode=block"',
```

and to disable headers from Magento output to allow the server to provide:

```
'x-frame-Options' => '_MAGE_DISABLED',
'x-content-type-options' => '_MAGE_DISABLED',
'x-xss-protection' => '_MAGE_DISABLED',
```

Also, I think it would be nice to consider including a configuration for these security headers in the Magento admin area, e.g. _Stores > Configuration > Advanced > Admin > Security_.  It could be something simple like 3 text field inputs, similar to the input fields for Lockout Time and Password Lifetime, with some preset system values, but with the option to override or disable if needed.

Thoughts anyone?

**Comment by maderlock** [comment]
@michaelstephenson states the use case pretty well:

If I want to embed certain pages in iframes then I want to be able to have more control over X-Frame-Options; most likely I'll output it for all but a few pages.

By having this output as a blanket setting I cannot do this, and thus cannot embed pages.

**Comment by orlangur** [comment]
```
/** Deployment config key for frontend x-frame-options header value */
const DEPLOYMENT_CONFIG_X_FRAME_OPT = 'x-frame-options';

/** Always send SAMEORIGIN in backend x-frame-options header */
const BACKEND_X_FRAME_OPT = 'SAMEORIGIN';
```

@maderlock are you saying that this comment is misleading and if you specify `'x-frame-options' => null` in `env.php` header is added for frontend?

> Luckily, we can handle this within the server settings. All we need to do is stop Magento from outputting this header itself.

What is the problem here if we can use `proxy_set_header` or `proxy_hide_header` in server configuration?

**Comment by maderlock** [comment]
I don't know what comment is supposed to be misleading, but yes, setting that value to null has no effect - the default is used.

Not everyone is using a reverse proxy, and even if they are you would have to ignore and replace the value rather than allow pages to set it as needed (not setting it as a blanket value)

**Comment by orlangur** [comment]
> I don't know what comment is supposed to be misleading, but yes, setting that value to null has no effect - the default is used.

On frontend? As far as I understand the comment `X-Frame-Options` must be always set for admin UI in current implementation.

> Not everyone is using a reverse proxy, and even if they are you would have to ignore and replace the value rather than allow pages to set it as needed (not setting it as a blanket value)

Is there some real-world scenario when it is hard to setup server in such a way that `X-Frame-Options` is filled with whatever you need, no matter what was returned by Magento? In the current implementation I really like that it is secure by default and there is no way to screw it up by accident.

**Comment by magento-engcom-team** [comment]
@maderlock, thank you for your report.
We've created internal ticket(s) MAGETWO-75571 to track progress on the issue.

**Comment by po2go-sean** [comment]
Is there any update on this issue? (Or any way to track the internal ticket?)

**Comment by ghost** [comment]
Hi guys any update on this?

**Comment by rip057** [comment]
I can report in magento 2.3.0-alpha that if you edit the env.php to say 
    'x-frame-options' => 'DENY',
a look at the headers section in firefox developer tools shows both
    X-Frame-Options: DENY
    X-Frame-Options: SAMEORIGIN
![githubxframe](https://user-images.githubusercontent.com/10294642/43637619-4be6b20a-96db-11e8-99db-b0ad65174126.jpg)

**Comment by kirkmadera** [comment]
Agreed that this should be able to be removed by setting it to null. Users may prefer to manage this at the server level. If Magento outputs this, it is not possible to manage server level because it ends up outputting the header twice.

**Comment by smgPatM** [comment]
Our team is fighting this same issue.   (All of this for older browsers that don't yet support the newer content-security-policy header.)

Yes, this should be more controllable in some way.
We don't want to globally allow ALL pages to be iframable.   A default XFO of SAMEORIGIN is fine.  But there has to be some way for application to modify/override the default XFO setting.
We need a programmatic way to defeat/override/replace the default XFO header and instead provide a customized x-frame-options ALLOW_FROM header.
(Or have default XFO value to NOT be injected into Response if some module/etc has already set an XFO header in Response object.)

Our use case is we have a few specific magento 2 paths/pages in a custom module that provide some store/product info that we want iframeable in pages on another domain.  But that ALLOWed 'parent' domainname will vary for each M2 store (and environment) and by path/page -- so a single global or even a single per-store setting won't suffice.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Echo 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-29754` were created

**Issue Available**: @engcom-Echo, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by magento-engcom-team** [comment]
Hi @maderlock. 

Thank you for your report and collaboration!

The related internal Jira ticket `MC-29754` was closed as `Fixed`. 

The fix will be available with the upcoming `2.4.3` release.
